<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Example 4: Using SEAGLE for Chromosome-Wide Gene-Based Analysis</title>

<script src="data:application/javascript;base64,Ly8gUGFuZG9jIDIuOSBhZGRzIGF0dHJpYnV0ZXMgb24gYm90aCBoZWFkZXIgYW5kIGRpdi4gV2UgcmVtb3ZlIHRoZSBmb3JtZXIgKHRvCi8vIGJlIGNvbXBhdGlibGUgd2l0aCB0aGUgYmVoYXZpb3Igb2YgUGFuZG9jIDwgMi44KS4KZG9jdW1lbnQuYWRkRXZlbnRMaXN0ZW5lcignRE9NQ29udGVudExvYWRlZCcsIGZ1bmN0aW9uKGUpIHsKICB2YXIgaHMgPSBkb2N1bWVudC5xdWVyeVNlbGVjdG9yQWxsKCJkaXYuc2VjdGlvbltjbGFzcyo9J2xldmVsJ10gPiA6Zmlyc3QtY2hpbGQiKTsKICB2YXIgaSwgaCwgYTsKICBmb3IgKGkgPSAwOyBpIDwgaHMubGVuZ3RoOyBpKyspIHsKICAgIGggPSBoc1tpXTsKICAgIGlmICghL15oWzEtNl0kL2kudGVzdChoLnRhZ05hbWUpKSBjb250aW51ZTsgIC8vIGl0IHNob3VsZCBiZSBhIGhlYWRlciBoMS1oNgogICAgYSA9IGguYXR0cmlidXRlczsKICAgIHdoaWxlIChhLmxlbmd0aCA+IDApIGgucmVtb3ZlQXR0cmlidXRlKGFbMF0ubmFtZSk7CiAgfQp9KTsK"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="data:text/css,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Example 4: Using SEAGLE for Chromosome-Wide Gene-Based Analysis</h1>



<p>This tutorial demonstrates how to use the <code>SEAGLE</code> package for chromosome-wide gene-based studies when the genotype data are from GWAS studies (.bed + .bim + .fam). We recommend using PLINK1.9 available from <a href="https://www.cog-genomics.org/plink/1.9/">https://www.cog-genomics.org/plink/1.9/</a> to generate a .raw file for each of the SNP sets. For more information on .raw files, please refer to <a href="https://www.cog-genomics.org/plink/2.0/formats#raw">https://www.cog-genomics.org/plink/2.0/formats#raw</a>. The .raw file records the allelic dosage for each SNP.</p>
<p>Examples files containing GWAS data (.bed, .bim, and .fam files) for all the genes in chromosome 22 and a corresponding gene list in <code>glist-hg19</code> can be downloaded via the <code>SEAGLE-example4-data.zip</code> file from <a href="http://jocelynchi.com/SEAGLE/SEAGLE-example4-data.zip">http://jocelynchi.com/SEAGLE/SEAGLE-example4-data.zip</a>.</p>
<p>To follow along with the PLINK conversion procedures, you will first need to install PLINK1.9 from <a href="https://www.cog-genomics.org/plink/1.9/">https://www.cog-genomics.org/plink/1.9/</a> and unzip the example <code>SEAGLE-example4-data.zip</code> file. Afterwards, you can type the PLINK commands below in a command prompt or terminal window in the folder where these files are located. This will produce a .raw file for each gene that you can read into R to obtain a relevant genetic marker matrix <span class="math inline">\({\bf G}\)</span> that you can input into <code>SEAGLE</code>.</p>
<p>After unzipping the <code>SEAGLE-example4-data.zip</code> file, you will find two empty directories named <code>plink_bash</code> and <code>R_codes</code>. The <code>plink_bash</code> directory is where we will write the PLINK1.9 commands that you will run to create the .raw files for each gene. The <code>R_codes</code> directory is where we will write the R scripts for loading the resulting .raw files into R.</p>
<div id="writing-.raw-files-for-each-gene-in-chromosome-22-with-plink1.9" class="section level2">
<h2>Writing .raw files for each gene in chromosome 22 with PLINK1.9</h2>
<p>We will begin by telling R the location where you unzipped the <code>SEAGLE-example4-data.zip</code> file. Then we will read in the gene list in <code>glist-hg19</code> and subset for the genes in chromosome 22. If you unzipped the file to your Downloads directory on a Mac, your path might look like the following.</p>
<pre><code>dir &lt;- &quot;/Users/user_name/Downloads/SEAGLE-example4-data/&quot;
glist &lt;- read.table(paste0(dir,&quot;glist-hg19&quot;), header = F)
glist_chr22 &lt;- subset(glist, glist$V1 == 22)</code></pre>
<p>Next, we will remove any duplicated genes for chromosome 22 in our gene list.</p>
<pre><code>dup &lt;- glist_chr22$V4[duplicated(glist_chr22$V4)] # find duplicated gene
dup</code></pre>
<p>Notice that in this example, there are 4 duplicated genes given by the following.</p>
<pre><code># GSTT2: chr22:24,322,339-24,326,106(GRCh37/hg19 by Ensembl)
# GSTT2B: chr22:24,299,601-24,303,373(GRCh37/hg19 by Ensembl)
# RIMBP3B: chr22:21,738,040-21,743,458(GRCh37/hg19 by Entrez Gene)
# RIMBP3C: chr22:21,899,646-21,905,750(GRCh37/hg19 by Ensembl)</code></pre>
<p>We will update our gene list by manually updating the positions of the duplicated genes.</p>
<pre><code>glist_chr22 &lt;- glist_chr22[!duplicated(glist_chr22$V4),]
glist_chr22[which(glist_chr22$V4 == &quot;GSTT2&quot;),2:3] &lt;- c(24322339, 24326106)
glist_chr22[which(glist_chr22$V4 == &quot;GSTT2B&quot;),2:3] &lt;- c(24299601,24303373)
glist_chr22[which(glist_chr22$V4 == &quot;RIMBP3B&quot;),2:3] &lt;- c(21738040,21743458)
glist_chr22[which(glist_chr22$V4 == &quot;RIMBP3C&quot;),2:3] &lt;- c(21899646,21905750)
write.table(glist_chr22, &quot;glist-hg19_chr22_updated&quot;, col.names = F, row.names = F, quote = F)

# replace hyphen with underscore
glist_chr22$V4 &lt;- sub(&quot;-&quot;, &quot;_&quot;, glist_chr22$V4)</code></pre>
<p>Now we will use PLINK1.9 to create a .raw file for each gene in chromosome 22, which includes all SNPs for a given gene. The following code will produce a sequence of bash files to produce .raw files for each gene in <code>glist_chr22</code>. It will also produce a corresponding sequence of R code snippets for reaching each .raw file into R. Each bash file will run PLINK1.9 for <code>num_genes=50</code> genes at a time.</p>
<pre><code># of plink commands per job
num_genes &lt;- 50

bash_plink &lt;- NULL # bash command
command &lt;- c(&quot;#! /bin/bash&quot;, &quot;#SBATCH --job-name=plink&quot;) # plink command
codeR &lt;- NULL # R code
for (i in 1:nrow(glist_chr22)) {
  command &lt;- c(command, paste0(&quot;plink --bfile 1kg_phase1_chr22 --make-set
  glist-hg19_chr22_updated --gene &quot;, glist_chr22$V4[i],&quot; --out &quot;,
  glist_chr22$V4[i],&quot; --export A&quot;) )

  # write R codes for reading .raw files, one R file for one gene
  codeR &lt;- paste0(glist_chr22$V4[i], &quot; &lt;- read.delim(\&quot;&quot;, glist_chr22$V4[i],&quot;.raw\&quot;, sep=\&quot; \&quot;)&quot;)
  fileConn&lt;-file(paste0(dir,&quot;R_codes/&quot;, glist_chr22$V4[i],&quot;.R&quot;))
  writeLines(codeR, fileConn)
  close(fileConn)

  if(i %% num_genes == 0 | i==nrow(glist_chr22)){
    # write plink commands in bash file
    fileConn&lt;-file(paste0(dir,&quot;plink_bash/plink_job&quot;,i-num_genes+1,&quot;-&quot;,i,&quot;.sh&quot;))
    writeLines(command, fileConn)
    close(fileConn)

    command &lt;- c(&quot;#! /bin/bash&quot;, &quot;#SBATCH --job-name=plink&quot;)
    bash_plink &lt;- c(bash_plink, paste0(&quot;sbatch plink_job&quot;,i-num_genes+1,&quot;-&quot;,i,&quot;.sh&quot;))
  }
}

# Write bash commands to a text file
fileConn&lt;-file(paste0(&quot;bash_plink.txt&quot;))
writeLines(bash_plink, fileConn)
close(fileConn)</code></pre>
<p>Note that if you want to run the .sh files in the <code>plink_bash</code> directory on your local computer, you can employ the <code>bash</code> command instead of <code>SBATCH</code> in terminal or command prompt. Additionally, if you are using Linux, you may need to specify <code>plink1.9</code> rather than just <code>plink</code>.</p>
<p>Running the above code will setup the code and files to produce a single .raw file for each gene when running the commands in the <code>bash_plink.txt</code> file. The .raw file for each gene will contain all the SNPs for that gene.</p>
</div>
<div id="loading-.raw-files-into-r-and-extracting-the-genetic-marker-matrix-bf-g" class="section level2">
<h2>Loading .raw files into R and extracting the genetic marker matrix <span class="math inline">\({\bf G}\)</span></h2>
<p>Now that we have data for each gene in the .raw format, we can load the .raw files sequentially into R for input into <code>SEAGLE</code> as follows. We’ll first load the <code>SEAGLE</code> package.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SEAGLE)</span></code></pre></div>
<p>The following code shows how to test for the GxE interaction effect using <code>SEAGLE</code> on these genes. To illustrate how to use SEAGLE, we will generate synthetic phenotype and covariate data for <span class="math inline">\(n=1092\)</span> study participants.</p>
<p>As an example, we’ve included .raw files for the following subset of genes from chromosome 22 in the <code>extdata</code> folder of this package: ACR, APOBEC3A, APOBEC3C, ARSA, ATF4, ATP5L2, BCRP2, BMS1P17, and BMS1P18. We’ve also included a corresponding gene list in <code>glist-hg19_chr22_example</code>. In practice, you will want to specify your own <code>dir</code> for the directory where you’ve stored your .raw files.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in gene list</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>dir <span class="ot">&lt;-</span> <span class="st">&quot;../inst/extdata/&quot;</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>genelist <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="fu">paste0</span>(dir, <span class="st">&quot;glist-hg19_chr22_example&quot;</span>), <span class="at">sep=</span><span class="st">&quot; &quot;</span>, <span class="at">header=</span><span class="cn">FALSE</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify number of genes in genelist</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>num_genes <span class="ot">&lt;-</span> <span class="fu">dim</span>(genelist)[<span class="dv">1</span>]</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate synthetic phenotype and covariate data</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">1092</span> <span class="co"># number of study participants</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> <span class="fu">rnorm</span>(n)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2</span>)</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">rnorm</span>(n))</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">3</span>)</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>E <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">rnorm</span>(n))</span></code></pre></div>
<p>Now that we have <span class="math inline">\({\bf y}\)</span>, <span class="math inline">\({\bf X}\)</span>, <span class="math inline">\({\bf E}\)</span>, and <span class="math inline">\({\bf G}\)</span>, we can run <code>SEAGLE</code> on each gene in <code>genelist</code>. We will first perform data checking procedures on <span class="math inline">\({\bf G}\)</span>. Then we will input <span class="math inline">\({\bf y}\)</span>, <span class="math inline">\({\bf X}\)</span>, <span class="math inline">\({\bf E}\)</span>, and <span class="math inline">\({\bf G}\)</span> for each gene into the <code>prep.SEAGLE</code> function. The <code>intercept = 0</code> parameter indicates that the first column of <span class="math inline">\({\bf X}\)</span> is not the all ones vector for the intercept.</p>
<p>The preparation procedure formats the input data for the <code>SEAGLE</code> function by checking the dimensions of the input data. It also pre-computes a QR decomposition for <span class="math inline">\(\widetilde{\bf X} = \begin{pmatrix} {\bf 1}_{n} &amp; {\bf X} &amp; {\bf E} \end{pmatrix}\)</span>, where <span class="math inline">\({\bf 1}_{n}\)</span> denotes the all ones vector of length <span class="math inline">\(n\)</span>.</p>
<p>Afterwards, we will input the prepared data for each gene into the <code>SEAGLE</code> function to compute the score-like test statistic <span class="math inline">\(T\)</span> and its corresponding p-value. The <code>init.tau</code> and <code>init.sigma</code> parameters are the initial values for estimating <span class="math inline">\(\tau\)</span> and <span class="math inline">\(\sigma\)</span> in the REML EM algorithm.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize output containers for T and p-value for each gene</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>T.list <span class="ot">&lt;-</span> <span class="fu">numeric</span>(<span class="at">length=</span>num_genes)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>pv.list <span class="ot">&lt;-</span> <span class="fu">numeric</span>(<span class="at">length=</span>num_genes)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Run SEAGLE on each gene in genelist</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>num_genes) {</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Identify current gene</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  gene_name <span class="ot">&lt;-</span> genelist[i,<span class="dv">4</span>]</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Obtain G</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  Gtemp <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="fu">paste0</span>(dir, gene_name, <span class="st">&quot;.raw&quot;</span>), <span class="at">sep=</span><span class="st">&quot; &quot;</span>)</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  G <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(Gtemp[,<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)])</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  L <span class="ot">&lt;-</span> <span class="fu">dim</span>(G)[<span class="dv">2</span>]                <span class="do">## number of SNPs</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Make weights</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  avg_newsnp <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(G)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>  fA  <span class="ot">=</span> avg_newsnp<span class="sc">/</span><span class="dv">2</span>            <span class="do">## freq of allele &quot;A&quot;</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>  fa  <span class="ot">=</span> <span class="dv">1</span><span class="sc">-</span>fA                    <span class="do">## freq of allele &quot;a&quot;</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>  maf <span class="ot">=</span> fA; maf[fA<span class="sc">&gt;</span><span class="fl">0.5</span>]<span class="ot">=</span> fa[fA<span class="sc">&gt;</span><span class="fl">0.5</span>]<span class="do">## maf should be b/w 0 and 0.5</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>  G <span class="ot">=</span> G[ ,maf<span class="sc">&gt;</span><span class="dv">0</span>]                <span class="do">## only keep those SNPs with MAF&gt;0</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>  maf <span class="ot">&lt;-</span> maf[maf <span class="sc">&gt;</span> <span class="dv">0</span>]           <span class="do">## Keep only MAF &gt; 0)</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>  wt   <span class="ot">=</span> <span class="fu">sqrt</span>(maf<span class="sc">^</span>(<span class="sc">-</span><span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>))       <span class="do">## Note we take the square root here</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">length</span>(wt) <span class="sc">&gt;</span> <span class="dv">1</span>) {</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>    G_final    <span class="ot">=</span> G <span class="sc">%*%</span> <span class="fu">diag</span>(wt) <span class="do">## Input this G</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>    T.list[i] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>    pv.list[i] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>    <span class="cf">next</span></span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Run SEAGLE</span></span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>  objSEAGLE <span class="ot">&lt;-</span> <span class="fu">prep.SEAGLE</span>(<span class="at">y=</span>y, <span class="at">X=</span>X, <span class="at">intercept=</span><span class="dv">0</span>, </span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>                           <span class="at">E=</span>E, <span class="at">G=</span>G_final)</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>  res <span class="ot">&lt;-</span> <span class="fu">SEAGLE</span>(objSEAGLE, <span class="at">init.tau=</span><span class="fl">0.5</span>, <span class="at">init.sigma=</span><span class="fl">0.5</span>)</span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Save T and p-value into output lists</span></span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>  T.list[i] <span class="ot">&lt;-</span> res<span class="sc">$</span>T</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a>  pv.list[i] <span class="ot">&lt;-</span> res<span class="sc">$</span>pv</span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>The score-like test statistics <span class="math inline">\(T\)</span> for the G<span class="math inline">\(\times\)</span>E effect and their corresponding p-values can be found in <code>T.list</code> and <code>pv.list</code>, respectively. We can take a look at the test statistics and p-values computed for each of the genes in <code>genelist</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>resMat <span class="ot">&lt;-</span> <span class="fu">cbind</span>(T.list, pv.list)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(resMat) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;p-value&quot;</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>resMat</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;               T   p-value</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1,] 7693.9187 0.6148780</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [2,] 4022.4392 0.7990406</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [3,] 3375.4804 0.9771967</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [4,] 9698.0070 0.2089241</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [5,] 3523.1752 0.3668638</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [6,]  716.1017 0.3435655</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [7,] 4792.4809 0.9336786</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [8,]  139.5526 0.9186002</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [9,]  139.5526 0.9186002</span></span></code></pre></div>
</div>
<div id="acknowledgments" class="section level2">
<h2>Acknowledgments</h2>
<p>Many thanks to Yueyang Huang for generating the example data, PLINK1.9 code and bash files, and R scripts for reading in the .raw files for this tutorial.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
